### DISCUSSION TITLE
LocalSocketListener security

### DISCUSSION THREAD

**Discussion by NuSkooler** [kotauskas/interprocess#55] [General]
I have implemented a PoC level pipe/UNIX domain socket system using `LocalSocketListener`. My next step is to secure the domain socket/pipes. Is this possible with this implementation or must I use the lower level OS-specific calls? I'd certainly prefer to have a generic code path with as little `target_os` checks as possible.

* Windows: Set the security descriptor
* Linux: File permissions

```rust
// how to secure ipc_listener?
let ipc_listener = LocalSocketListener::bind(local_ipc_name.clone())?;
```

**Comment by kotauskas** [comment]
[On Linux](<https://www.man7.org/linux/man-pages/man7/unix.7.html>), the created socket file obeys usual [`umask()`](<https://www.man7.org/linux/man-pages/man2/umask.2.html>) logic. None of the calls performed by Interprocess take the mode as an input. On Windows, `PipeListenerOptions` on named pipes allows using a custom security descriptor (this requires unsafe code, and I might overhaul it a bit in the following commits).

Adding a `local_socket::ListenerOptions` that reads certain parameters and ignores ones not applicable to the current system seems like the most ergonomic choice to me. I haven't considered it before, but permissions sound like good enough motivation for it to be a thing.

**Comment by NuSkooler** [comment]
@kotauskas Thanks for your quick reply! `ListenerOptions` on `local_socket` sounds great. Any way I can help?

**Reply by kotauskas** [reply]
Because the APIs are currently in rapid motion in preparation for 2.0.0 (I'm consolidating all previously planned breaking changes), now is an inopportune time for PRs. This should be over in a few weeks depending on how much free time I'll get to spend on this.

**Reply by NuSkooler** [reply]
Thanks, I'll keep an eye out - what's the 2.0 ETA?

**Reply by kotauskas** [reply]
@NuSkooler 2.0.0 is now out (in fact, 2.1.0 is the most recent version), and both Windows security descriptors and Unix file modes are now supported.

**Reply by NuSkooler** [reply]
Great, thank you!
